examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1212 @@
+// function, but we separately time the set-up of the preconditioner as well as the GMRES solve.
+  template <int dim>
+  void StokesProblem<dim>::solve ()
+  {
+    SolverControl solver_control (system_matrix.m(),
+                                  1e-6*system_rhs.l2_norm());
+
+    GrowingVectorMemory<BlockVector<double> > vector_memory;
+    SolverGMRES<BlockVector<double> >::AdditionalData gmres_data;
+    gmres_data.max_n_tmp_vectors = 100;
+
+    SolverGMRES<BlockVector<double> > gmres(solver_control, vector_memory,
+                                            gmres_data);
+
+
+    if (use_multigrid==false)
+      {
+
+        computing_timer.enter_subsection ("Solve - Set-up Preconditioner");
+
+        std::cout << "   Computing preconditioner..." << std::endl << std::flush;
+
+        SparseILU<double>       A_preconditioner;
+        A_preconditioner.initialize (system_matrix.block(0,0),SparseILU<double>::AdditionalData());
+
+
+        SparseMatrix<double> pressure_mass_matrix;
+        pressure_mass_matrix.reinit(sparsity_pattern.block(1,1));
+        pressure_mass_matrix.copy_from(system_matrix.block(1,1));
+        system_matrix.block(1,1) = 0;
+
+
+        SparseILU<double> pmass_preconditioner;
+        pmass_preconditioner.initialize (pressure_mass_matrix,
+                                         SparseILU<double>::AdditionalData());
+
+        bool use_expensive = false;
+        unsigned int its_A = 0, its_S = 0;
+        // If this cheaper solver is not desired, then simply short-cut
+        // the attempt at solving with the cheaper preconditioner that consists
+        // of only a single V-cycle
+        const BlockSchurPreconditioner<typename InnerPreconditioner<dim>::type,
+              SparseILU<double>>
+              preconditioner (system_matrix,
+                              pressure_mass_matrix,
+                              pmass_preconditioner, A_preconditioner,
+                              use_expensive);
+        computing_timer.leave_subsection();
+        computing_timer.enter_subsection ("Solve - GMRES");
+        gmres.solve (system_matrix,
+                     solution,
+                     system_rhs,
+                     preconditioner);
+        computing_timer.leave_subsection();
+
+        its_A += preconditioner.n_iterations_A();
+        its_S += preconditioner.n_iterations_S();
+
+        constraints.distribute (solution);
+
+        std::cout << " "
+                  << solver_control.last_step()
+                  << " block GMRES iterations";
+      }
+    else
+      {
+        computing_timer.enter_subsection ("Solve - Set-up Preconditioner");
+        // Transfer operators between levels
+        MGTransferPrebuilt<Vector<double> > mg_transfer(hanging_node_constraints, mg_constrained_dofs);
+        mg_transfer.build_matrices(velocity_dof_handler);
+
+        // Coarse grid solver
+        FullMatrix<double> coarse_matrix;
+        coarse_matrix.copy_from (mg_matrices[0]);
```

Comment:
investigate if we can get rid of this copy and just do ``coarse_grid_solver.initialize(mg_matrices[0])``